homeassistant/components/eq3btsmart/climate.py

```diff
@@ -0,0 +1,320 @@
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.helpers.device_registry import (
+    CONNECTION_BLUETOOTH,
+    DeviceInfo,
+    async_get,
+    format_mac,
+)
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import (
+    DEVICE_MODEL,
+    DOMAIN,
+    EQ_TO_HA_HVAC,
+    HA_TO_EQ_HVAC,
+    MANUFACTURER,
+    SIGNAL_THERMOSTAT_CONNECTED,
+    SIGNAL_THERMOSTAT_DISCONNECTED,
+    CurrentTemperatureSelector,
+    Preset,
+    TargetTemperatureSelector,
+)
+from .entity import Entity
+from .models import Eq3Config, Eq3ConfigEntryData
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Handle config entry setup."""
+
+    eq3_config_entry: Eq3ConfigEntryData = hass.data[DOMAIN][config_entry.entry_id]
+
+    async_add_entities(
+        [Eq3Climate(eq3_config_entry.eq3_config, eq3_config_entry.thermostat)],
+    )
+
+
+class Eq3Climate(Entity, ClimateEntity):
+    """Climate entity to represent a eQ-3 thermostat."""
+
+    _attr_has_entity_name: bool = True
+    _attr_name: str | None = None
+    _attr_supported_features: ClimateEntityFeature = (
+        ClimateEntityFeature.TARGET_TEMPERATURE
+        | ClimateEntityFeature.PRESET_MODE
+        | ClimateEntityFeature.TURN_OFF
+        | ClimateEntityFeature.TURN_ON
+    )
+    _attr_temperature_unit: UnitOfTemperature = UnitOfTemperature.CELSIUS
+    _attr_min_temp: float = EQ3BT_OFF_TEMP
+    _attr_max_temp: float = EQ3BT_MAX_TEMP
+    _attr_precision: float = PRECISION_HALVES
+    _attr_hvac_modes: list[HVACMode] = list(HA_TO_EQ_HVAC.keys())
+    _attr_preset_modes: list[str] = list(Preset)
+    _attr_should_poll: bool = False
+
+    def __init__(self, eq3_config: Eq3Config, thermostat: Thermostat) -> None:
+        """Initialize the climate entity."""
+
+        super().__init__(eq3_config, thermostat)
+
+        self._target_temperature: float | None = None
+        self._attr_available = False
+        self._attr_hvac_mode: HVACMode | None = None
+        self._attr_hvac_action: HVACAction | None = None
+        self._attr_preset_mode: str | None = None
+        self._attr_unique_id = format_mac(self._eq3_config.mac_address)
+        self._attr_device_info: DeviceInfo = DeviceInfo(
+            name=self._eq3_config.name,
+            manufacturer=MANUFACTURER,
+            model=DEVICE_MODEL,
+            sw_version=None,
+            serial_number=None,
+            connections={(CONNECTION_BLUETOOTH, self._eq3_config.mac_address)},
+        )
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+
+        self._thermostat.register_update_callback(self._async_on_updated)
```

Comment:
How does this get removed when the integration is unloaded/reloaded?

Reply:
It does not, I thought the `Thermostat` object gets recreated when the integration is reloaded, but that's obviously not the case. I'll implement a `unregister_update_callback` in the library and run that in `async_will_remove_from_hass`.

Reply:
I think `register_connection_callback` probably has the same issue?

Reply:
It had, but it's not used anymore since I handle the connection and reconnection logic in the integration now.